### PR TITLE
Citadel generates the istiod certificate

### PR DESCRIPTION
**Pull request by lei-tang**
Please provide a description for what this PR is for: https://github.com/istio/istio/issues/19950

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by howardjohn** [comment]
Don't we need changes to pilot-agent to support this as well?

**Comment by lei-tang** [comment]
@howardjohn Yes, we also need changes on pilot-agent to support this as well, which will be a separate PR to make this PR smaller and easier to review.

**Comment by howardjohn** [comment]
> PR smaller and easier to review.

SGTM

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -138,7 +141,7 @@ type Server struct {
 }
 
 // NewServer creates a new Server instance based on the provided arguments.
-func NewServer(args *PilotArgs) (*Server, error) {
+func NewServer(args *PilotArgs, stopCh <-chan struct{}) (*Server, error) {
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/server.go]: This is not right. stopCh is never actually used. I think we need to use the same stopCh that is used for startFunc which may mean we do everything inside startFuncs

Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/server.go]: Yes, basically the stop channel in the Server is only for test.

File: pilot/pkg/bootstrap/certcontroller.go [hunk 2]
```diff
@@ -22,6 +22,8 @@ import (
 	"strings"
 	"time"
 
+	"istio.io/istio/pilot/pkg/features"
+
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/certcontroller.go]: remove

File: pilot/pkg/bootstrap/certcontroller.go [hunk 3]
```diff
@@ -37,6 +39,9 @@ const (
 	// Default CA certificate path
 	// Currently, custom CA path is not supported; no API to get custom CA cert yet.
 	defaultCACertPath = "./var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+
+	// pilotCertTTL is the TTL of Pilot certificate signed by Citadel.
+	pilotCertTTL = 3650 * 24 * time.Hour
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/certcontroller.go]: Can we reuse the `selfSignedCACertTTL`?

File: pilot/pkg/bootstrap/certcontroller.go [hunk 4]
```diff
@@ -129,11 +134,16 @@ func (s *Server) initDNSCerts(hostname string) error {
...
 
-	log.Infoa("Generating K8S-signed cert for ", names)
-
-	// TODO: fallback to citadel (or custom CA) if K8S signing is broken
-	certChain, keyPEM, _, err := chiron.GenKeyCertK8sCA(s.kubeClient.CertificatesV1beta1().CertificateSigningRequests(),
-		strings.Join(names, ","), parts[0]+".csr.secret", parts[1], defaultCACertPath)
+	var certChain, keyPEM []byte
+	var err error
+	if features.SignCertAtK8sCa.Get() {
+		log.Infoa("Generating K8S-signed cert for ", names)
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/certcontroller.go]: Infof %v

File: pilot/pkg/bootstrap/server.go [hunk 5]
```diff
@@ -27,6 +27,8 @@ import (
 	"strings"
 	"time"
 
+	"istio.io/istio/security/pkg/pki/ca"
+
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go]: blank

File: pilot/pkg/bootstrap/certcontroller.go (hunk 2 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/certcontroller.go]: features package is used on the line "if features.SignCertAtK8sCa.Get() {"

Comment (lei-tang) [reply pilot/pkg/bootstrap/certcontroller.go]: The blank line under "istio.io/istio/pilot/pkg/features" has been removed.

File: pilot/pkg/bootstrap/certcontroller.go (hunk 3 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/certcontroller.go]: Done.

File: pilot/pkg/bootstrap/certcontroller.go (hunk 4 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/certcontroller.go]: Done.

File: pilot/pkg/bootstrap/server.go (hunk 5 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/server.go]: The blank line has been removed.

File: pkg/test/framework/components/pilot/native.go [hunk 6]
```diff
@@ -127,8 +127,9 @@ func newNative(ctx resource.Context, cfg Config) (Instance, error) {
 	bootstrapArgs.MCPMaxMessageSize = 1024 * 1024 * 4
 
 	var err error
+	stopCh := make(chan struct{})
 	// Create the server for the discovery service.
-	if instance.server, err = bootstrap.NewServer(&bootstrapArgs); err != nil {
+	if instance.server, err = bootstrap.NewServer(&bootstrapArgs, stopCh); err != nil {
```
Comment (howardjohn) [inline pkg/test/framework/components/pilot/native.go]: fix stop here

Comment (lei-tang) [reply pkg/test/framework/components/pilot/native.go]: Fixed. Thanks!

File: pilot/pkg/bootstrap/server.go (hunk 1 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/server.go]: Per discussion, the fix has been added to pkg/test/framework/components/pilot/native.go.

**Pull request merged**